File: Week_06/G20190343020039/LeetCode_200_0039.py
class Solution:

  def numIslands(self, grid: List[List[str]]) -> int:
    if not grid: return 0
    rows = len(grid)
    cols = len(grid[0])
    p = {}
    for row in range(rows):
      for col in range(cols):
        if grid[row][col] == "1":
          p[row * cols + col] = row * cols + col

    directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]
    for row in range(rows):
      for col in range(cols):
        if grid[row][col] == "1":
          for _, dirs in enumerate(directions):
            newRow = row + dirs[0]
            newCol = col + dirs[1]
            if self.isValid(grid, newRow, newCol):
              self.union(p, row * cols + col, newRow * cols + newCol)

    return len(set([self.parent(p, key) for key in p]))

  def isValid(self, grid, row, col):
    if row < 0 or row >= len(grid):
      return False
    if col < 0 or col >= len(grid[0]):
      return False
    if grid[row][col] == "0":
      return False
    return True

  def union(self, p, i, j):
    p1 = self.parent(p, i)
    p2 = self.parent(p, j)
    p[p1] = p2

  def parent(self, p, i):
    root = i
    while p[root] != root:
      root = p[root]
    while p[i] != i:
      x = i
      i = p[i]
      p[x] = root
    return root

File: Week_03/G20190343020039/LeetCode_200_0039.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
  if (grid.length === 0) {
    return 0;
  }
  let rows = grid.length;
  let cols = grid[0].length;
  let count = 0;
  for (let i = 0; i < rows; i += 1) {
    for (let j = 0; j < cols; j += 1) {
      if (grid[i][j] === "1") {
        count += 1;
        destroyIsland(grid, i, j);
      }
    }
  }
  return count;
};

function destroyIsland(grid, row, col) {
  const dirs = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1]
  ];
  grid[row][col] = "0";
  let queue = [];
  queue.push([row, col]);
  while (queue.length > 0) {
    let cur = queue.shift();
    let curRow = cur[0];
    let curCol = cur[1];
    for (let i = 0; i < dirs.length; i += 1) {
      let newRow = curRow + dirs[i][0];
      let newCol = curCol + dirs[i][1];
      if (isIndexValid(grid, newRow, newCol) &&
        grid[newRow][newCol] === "1") {
        grid[newRow][newCol] = "0";
        queue.push([newRow, newCol]);
      }
    }
  }
}

function isIndexValid(grid, row, col) {
  if (row < 0 || col < 0) {
    return false;
  }
  if (row >= grid.length) {
    return false;
  }
  if (col >= grid[0].length) {
    return false;
  }
  return true;
}
